kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/MongoDebeziumTransform.java

```diff
@@ -0,0 +1,334 @@
+
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.apache.kafka.connect.transforms.ExtractField;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.apache.kafka.connect.transforms.util.Requirements;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Debezium Mongo Connector generates the CDC before/after fields as BSON strings. This SMT converts
+ * those strings into typed SinkRecord Structs by inferring the schema from the BSON node types.
+ *
+ * @param <R> the subtype of {@link ConnectRecord} on which this transformation will operate, most
+ *     likely SinkRecord
+ */
+public class MongoDebeziumTransform<R extends ConnectRecord<R>> implements Transformation<R> {
+
+  public static final String ARRAY_HANDLING_MODE_KEY = "array_handling_mode";
+
+  public static final String RECORD_ENVELOPE_KEY_SCHEMA_NAME_SUFFIX = ".Key";
+
+  public static final String SCHEMA_NAME_SUFFIX = ".Envelope";
+
+  private static final String UPDATE_DESCRIPTION = "updateDescription";
+
+  private static final String REMOVED_FIELDS = "removedFields";
+  private static final String UPDATED_FIELDS = "updatedFields";
+
+  private static final String AFTER_FIELD_NAME = "after";
+  private static final String BEFORE_FIELD_NAME = "before";
+  private final ExtractField<R> updateDescriptionExtractor =
+      extractorValueField(UPDATE_DESCRIPTION);
+
+  private final ExtractField<R> afterExtractor = extractorValueField(AFTER_FIELD_NAME);
+
+  private final ExtractField<R> beforeExtractor = extractorValueField(BEFORE_FIELD_NAME);
+
+  private final ExtractField<R> keyIdExtractor = extractorForKeyField("id");
+
+  private MongoDataConverter converter;
+
+  private static final Logger LOG = LoggerFactory.getLogger(MongoDebeziumTransform.class.getName());
+
+  private static final ConfigDef CONFIG_DEF =
+      new ConfigDef()
+          .define(
+              ARRAY_HANDLING_MODE_KEY,
+              ConfigDef.Type.STRING,
+              null,
+              ConfigDef.Importance.MEDIUM,
+              "array or document handling for mongodb arrays");
+
+  @Override
+  public R apply(R record) {
+    // pass tombstones as-is
+    if (record.value() == null) {
+      return record;
+    }
+
+    // if they don't contain key/envelope convert to tombstone
+    if (!isValidKey(record) || !isValidValue(record)) {
+      return record.newRecord(
+          record.topic(),
+          record.kafkaPartition(),
+          null,
+          null,
+          null,
+          null,
+          record.timestamp(),
+          record.headers());
+    }
```

Comment:
and then over here I would just do something like: 

```suggestion
  private SinkRecord tombstone(SinkRecord record) {
    return record.newRecord(
            record.topic(),
            record.kafkaPartition(),
            null,
            null,
            null,
            null,
            record.timestamp(),
            record.headers());
  }

    if (!isValidKey(record)) {
      LOG.debug("Expected Key Schema for transformation, converting to tombstone. Message key: \"{}\"", record.key());
      return tombstone(record);
    }

    if (!isValidValue(record)) {
      LOG.debug("Expected Envelope for transformation, converting to tombstone");
      return tombstone(record);
    }
```

You may also want to make the log lines more consistent between the two.